### PR TITLE
Backend-driven Dynamic Cards filtering

### LINKED ISSUE #22315: 🍎 Add backend-driven Dynamic Cards filtering
- Remove the old filtering logic that was performed 
- Pass the following params to the dashboard cards endpoint:
  - device_id
  - platform
  - build_number
  - marketing_version
  - identifier 
 
Ref p1704176298626839-slack-C067H36F011

### PR DESCRIPTION
**Pull request by salimbraksa**
Fixes #22315

## Related PR
- https://github.com/wordpress-mobile/WordPressKit-iOS/pull/674

## Descrption
This PR calls the updates dashboard cards endpoint passing the extra parameters (device_id, platform, build_number, marketing_version, identifier) and removes the local dynamic cards filtering logic that was based on the deleted `remote_feature_flag` response field.

## Test Instructions
- Sandbox your emulator with D133371-code (instructions paqN3M-5m-p2)
- Open `FeatureFlagRemote.swift` file.
- At line 86, change `self.marketingVersion` to `23.9`.
- **Verify** that the domain management card does not appear
- Open `FeatureFlagRemote.swift` file.
- At line 86, change `self.marketingVersion` to `24.0`.
- **Verify** that the domain management card appears

## Note 
Please remove the `Do Not Merge` label and merge after:
- the backend change is deployed D133371-code
- The related WordPressKit PR is merged.
- The `WordPressKit` pod on this PR is updated to point to `trunk`.

## Regression Notes
1. Potential unintended areas of impact
New params are passed to the `Dashboard Cards` endpoint, this API change could unintentionally return incorrect cards. So, the content of the following cards should be tested:
   1. Today's Stats
   2. Scheduled Posts
   3. Activity
   4. Drafts
   5. Pages

2. What I did to test those areas of impact (or what existing automated tests I relied on)
Manual testing.

3. What automated tests I added (or what prevented me from doing so)
None.

## PR submission checklist:
- [x] I have completed the Regression Notes.
- [x] I have considered adding unit tests for my changes.
- [x] I have considered adding accessibility improvements for my changes.
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

### CODE REVIEW

File: Podfile [hunk 1]
```diff
@@ -52,8 +52,8 @@ end
 def wordpress_kit
   # Anything compatible with 8.9, starting from 8.9.1 which has a breaking change fix
   # pod 'WordPressKit', '~> 9.0'
-  pod 'WordPressKit', git: 'https://github.com/wordpress-mobile/WordPressKit-iOS.git', commit: '32b5a9fdd097b82934a5ce679c915a1fc3f21848'
-  # pod 'WordPressKit', git: 'https://github.com/wordpress-mobile/WordPressKit-iOS.git', branch: 'trunk'
+  # pod 'WordPressKit', git: 'https://github.com/wordpress-mobile/WordPressKit-iOS.git', commit: '32b5a9fdd097b82934a5ce679c915a1fc3f21848'
+  pod 'WordPressKit', git: 'https://github.com/wordpress-mobile/WordPressKit-iOS.git', branch: 'task/22315-backend-driven-dynamic-cards-filtering'
```
Comment (hassaanelgarem) [inline Podfile]: I had to do this to get the rake dependencies to succeed:
```suggestion
  pod 'WordPressKit', git: 'https://github.com/wordpress-mobile/WordPressKit-iOS.git', commit: 'accfc22fada868a8b44247acaf15c9edb403ca43'
  # pod 'WordPressKit', git: 'https://github.com/wordpress-mobile/WordPressKit-iOS.git', branch: ''
```

**Review by hassaanelgarem — APPROVED** [review]
Tested against the sandbox and it works as expected 🚀 

> Sandbox your emulator with [D133371-code](https://href.li/?https://code.a8c.com/D133371) (instructions [paqN3M-5m-p2](https://href.li/?https://wp.me/paqN3M-5m))

FYI: the instructions linked here only applied to Android. For iOS, I was able to test by simply sandboxing `public-api.wordpress.com` on my mac and running a simulator.

File: WordPress/Classes/Stores/RemoteFeatureFlagStore.swift [hunk 2]
```diff
@@ -17,6 +17,19 @@ class RemoteFeatureFlagStore {
         return queue
       }()
 
+    /// The `deviceID` ensures we retain a stable set of Feature Flags between updates. If there are staged rollouts or other dynamic changes
+    /// happening server-side we don't want out flags to change on each fetch, so we provide an anonymous ID to manage this.
```
Comment (alpavanoglu) [inline WordPress/Classes/Stores/RemoteFeatureFlagStore.swift]: typo: "out flags" -> "our flags" I suppose?
But I think you've only moved this code.

**Review by alpavanoglu — APPROVED** [review]
Code looks good. I did not test since Hassaan has already approved. LMK if you prefer that I test as well.

File: WordPress/Classes/Stores/RemoteFeatureFlagStore.swift (hunk 2 above)
Comment (salimbraksa) [reply WordPress/Classes/Stores/RemoteFeatureFlagStore.swift]: > But I think you've only moved this code.

Correct, but I'll fix the typo.

**Comment by salimbraksa** [comment]
Hey @mokagio, following up on this [comment](https://github.com/wordpress-mobile/WordPressKit-iOS/pull/674#issuecomment-1880247838), I've pinged @dvdchr to look into the flaky test.

**Comment by hassaanelgarem** [comment]
Hey @salimbraksa 👋 
6a7a1e397a392534fb74d97daf6af3791a94a53f should fix the failing test. I pushed directly to the branch in an attempt to catch the code freeze 🤞

**Comment by mokagio** [comment]
Hey @salimbraksa  👋 You might notice a merge `trunk` commit here from me. That is to bring this branch up to date with the latest CI changes on `trunk` and ensure everything works on the new Xcode 15.1 CI stack.

Don't hesitate to ping me if there's any trouble.

**Comment by salimbraksa** [comment]
Hey @mokagio, I've switched the base branch to `release/24.0` and everything's checked out. Could you take care of its release along with the `WordPressKit` PR? Thanks in advance! 🙇

**Comment by mokagio** [comment]
Hey @salimbraksa 👋 I'm always happy to help 👍 

https://github.com/wordpress-mobile/WordPressKit-iOS/pull/674 introduced a breaking change. That's okay, but [as I mentioned](https://github.com/wordpress-mobile/WordPressKit-iOS/pull/674#issuecomment-1880241165) is something that adds extra release overhead because [WordPressAuthenticator depends on WordPressKit](https://github.com/wordpress-mobile/WordPressAuthenticator-iOS/blob/0caf8f964a4604e53df16da5f2e9f209f91dc8aa/WordPressAuthenticator.podspec#L43). If we bump the major version of Kit, we also need to release a new version of Authenticator.

That's not a big deal, but given the additional effort, we try to minimize breaking changes and use the occasion when we _have_ to ship one as a chance to batch in others.

Is there any way the work from https://github.com/wordpress-mobile/WordPressKit-iOS/pull/674 could be updated so that it's backward compatible?

For example. The current diff goes:

```diff
 open func fetch(
         cards: [String],
         forBlogID blogID: Int,
+        deviceId: String,
         success: @escaping (NSDictionary) -> Void,
         failure: @escaping (Error) -> Void
     ) {
```

Could we do something like


```swift
open func fetch(
        cards: [String],
        forBlogID blogID: Int,
        deviceId: String? = nil,
        success: @escaping (NSDictionary) -> Void,
        failure: @escaping (Error) -> Void
     ) {
         guard let deviceId else {
            // A deviceId value must be provided, we allow it to be nil only to avoid breaking API compatibility.
            failure(NSError(domain: "DashboardServiceRemote", code: 1, userInfo: nil))
```

Admittedly, it's not an elegant solution and it leaves the door open for future misuses, but it does allow us to be more intentional with breaking changes.

What do you think?

**Comment by mokagio** [comment]
@salimbraksa thank you for quickly addressing my suggestion via https://github.com/wordpress-mobile/WordPressKit-iOS/pull/676 – And love that there was a test for that path, too.

I'll take it from here to release the new WordPressKit and integrate — _seems only fair_ 😄

**Comment by mokagio** [comment]
Taking over merging to ship a new build.

**Pull request merged**